presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java

```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
-        // check range validity
-        checkState(source != null, "source is null");
-        checkPositionIndexes(startPosition, startPosition + length, source.getPositionCount());
-
-        if (usingCopiedBlock) {
-            for (int i = 0; i < length; i++) {
-                type.appendTo(source, startPosition + i, outputBlockBuilder);
+        int nullIndex = nullFinder.getNullElementIndex();
+        verify(nullIndex != INVALID_NULL_POSITION, "nullIndex is -1");
+
+        int position = 0;
+        int[] ids = new int[maxTotalEntriesForBatch];
+        for (int i = 0; i < length; i++) {
+            int curEntries = entryCounts[offset + i];
+            for (int j = 0; j < curEntries; j++) {
+                ids[position++] = sourcePosition++;
             }
-        }
-        else {
-            for (int i = 0; i < length; i++) {
-                appendId(startPosition + i);
+
+            int curMaxEntries = maxEntries[offset + i];
+            for (int j = curEntries; j < curMaxEntries; j++) {
+                ids[position++] = nullIndex;
             }
         }
+
+        return new DictionaryBlock(maxTotalEntriesForBatch, source, ids);
     }
 
-    public void appendNull()
+    protected Block getSource()
     {
-        if (usingCopiedBlock) {
-            outputBlockBuilder.appendNull();
-        }
-        else {
-            int nullIndex = nullFinder.getNullElementIndex();
-            if (nullIndex == -1) {
-                startUsingCopiedBlock();
-                appendNull();
-            }
-            else {
-                appendId(nullIndex);
-            }
-        }
+        return source;
     }
 
-    /**
-     * This method is invoked when transition from dictionary block to copied block happens.
-     * - outputBlockBuilder is created and existing elements as indicated by ids[] are copied over.
-     * - ids[] and positionCount are reset, as they are not used anymore.
-     */
-    private void startUsingCopiedBlock()
+    protected int[] getEntryCounts()
     {
-        requireNonNull(pageBuilderStatus, "pageBuilderStatus is null");
-        // initialize state for copied block
-        outputBlockBuilder = type.createBlockBuilder(pageBuilderStatus.createBlockBuilderStatus(), ids.length);
-
-        // Transfer elements from ids[] to output page builder
-        for (int i = 0; i < positionCount; i++) {
-            type.appendTo(source, ids[i], outputBlockBuilder);
-        }
-
-        // reset state for dictionary block
-        ids = new int[0];
-        positionCount = 0;
-
-        usingCopiedBlock = true;
+        return entryCounts;
     }
 
-    /**
-     * This method appends {@code sourcePosition} to ids array.
-     *
-     * The caller of this method has to ensure the following:
-     * - {@code sourcePosition} is a valid index in the source
-     * - {@code usingCopiedBlock} is false
-     */
-    private void appendId(int sourcePosition)
+    protected NullElementFinder getNullFinder()
```

Comment:
This method is not needed. Just add `nullIndex` parameter to `buildOutputBlockWithNullsByDictionary`.